// angular-bootstrap-md/index.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
// free
var core_1 = require("@angular/core");
var _1 = require("./inputs/");
var buttons_1 = require("./buttons");
var ripple_1 = require("./ripple");
var inputs_1 = require("./inputs");
var navbars_1 = require("./navbars");
var dropdown_1 = require("./dropdown");
var _2 = require("./carousel/");
var _3 = require("./charts/");
var collapse_1 = require("./collapse");
var modals_1 = require("./modals");
var tooltip_1 = require("./tooltip");
var popover_1 = require("./popover");
var waves_module_1 = require("./waves/waves.module");
var _4 = require("./buttons/");
exports.ButtonsModule = _4.ButtonsModule;
exports.ButtonRadioDirective = _4.ButtonRadioDirective;
exports.ButtonCheckboxDirective = _4.ButtonCheckboxDirective;
var _5 = require("./ripple/");
exports.RippleModule = _5.RippleModule;
exports.RippleDirective = _5.RippleDirective;
var _6 = require("./waves/");
exports.WavesModule = _6.WavesModule;
exports.WavesDirective = _6.WavesDirective;
var _7 = require("./inputs/");
exports.DeepModule = _7.DeepModule;
exports.DeepDirective = _7.DeepDirective;
var inputs_2 = require("./inputs");
exports.InputsModule = inputs_2.InputsModule;
exports.MdbInputDirective = inputs_2.MdbInputDirective;
var _8 = require("./inputs/");
exports.ActiveModule = _8.ActiveModule;
exports.ActiveDirective = _8.ActiveDirective;
exports.InputValidateDirective = _8.InputValidateDirective;
var _9 = require("./navbars/");
exports.NavbarModule = _9.NavbarModule;
var _10 = require("./dropdown/");
exports.BsDropdownConfig = _10.BsDropdownConfig;
exports.BsDropdownContainerComponent = _10.BsDropdownContainerComponent;
exports.BsDropdownDirective = _10.BsDropdownDirective;
exports.BsDropdownMenuDirective = _10.BsDropdownMenuDirective;
exports.BsDropdownModule = _10.BsDropdownModule;
exports.BsDropdownState = _10.BsDropdownState;
exports.BsDropdownToggleDirective = _10.BsDropdownToggleDirective;
var _11 = require("./carousel/");
exports.CarouselComponent = _11.CarouselComponent;
exports.CarouselConfig = _11.CarouselConfig;
exports.CarouselModule = _11.CarouselModule;
var _12 = require("./charts/");
exports.MDBChartsModule = _12.MDBChartsModule;
exports.BaseChartDirective = _12.BaseChartDirective;
var _13 = require("./collapse/");
exports.CollapseDirective = _13.CollapseDirective;
exports.CollapseModule = _13.CollapseModule;
var _14 = require("./modals/");
exports.ModalBackdropComponent = _14.ModalBackdropComponent;
exports.ModalBackdropOptions = _14.ModalBackdropOptions;
exports.ModalDirective = _14.ModalDirective;
exports.ModalModule = _14.ModalModule;
exports.ModalOptions = _14.ModalOptions;
exports.MDBModalService = _14.MDBModalService;
exports.ModalContainerComponent = _14.ModalContainerComponent;
exports.MDBModalRef = _14.MDBModalRef;
var _15 = require("./tooltip/");
exports.TooltipConfig = _15.TooltipConfig;
exports.TooltipContainerComponent = _15.TooltipContainerComponent;
exports.TooltipDirective = _15.TooltipDirective;
exports.MDBTooltipModule = _15.MDBTooltipModule;
var _16 = require("./popover/");
exports.PopoverConfig = _16.PopoverConfig;
exports.PopoverContainerComponent = _16.PopoverContainerComponent;
exports.PopoverModule = _16.PopoverModule;
exports.PopoverDirective = _16.PopoverDirective;
var MODULES = [
    _1.DeepModule,
    buttons_1.ButtonsModule,
    ripple_1.RippleModule,
    waves_module_1.WavesModule,
    inputs_1.InputsModule,
    inputs_1.ActiveModule,
    navbars_1.NavbarModule,
    dropdown_1.BsDropdownModule,
    _2.CarouselModule,
    _3.MDBChartsModule,
    collapse_1.CollapseModule,
    modals_1.ModalModule,
    tooltip_1.MDBTooltipModule,
    popover_1.PopoverModule,
];
var MDBRootModule = /** @class */ (function () {
    function MDBRootModule() {
    }
    MDBRootModule = __decorate([
        core_1.NgModule({
            imports: [
                buttons_1.ButtonsModule,
                _1.DeepModule,
                ripple_1.RippleModule.forRoot(),
                waves_module_1.WavesModule.forRoot(),
                inputs_1.InputsModule.forRoot(),
                inputs_1.ActiveModule.forRoot(),
                navbars_1.NavbarModule,
                dropdown_1.BsDropdownModule.forRoot(),
                _2.CarouselModule.forRoot(),
                _3.MDBChartsModule,
                collapse_1.CollapseModule.forRoot(),
                modals_1.ModalModule.forRoot(),
                tooltip_1.MDBTooltipModule.forRoot(),
                popover_1.PopoverModule.forRoot(),
            ],
            exports: MODULES,
            schemas: [core_1.NO_ERRORS_SCHEMA]
        })
    ], MDBRootModule);
    return MDBRootModule;
}());
exports.MDBRootModule = MDBRootModule;
var MDBBootstrapModule = /** @class */ (function () {
    function MDBBootstrapModule() {
    }
    MDBBootstrapModule.forRoot = function () {
        return { ngModule: MDBRootModule };
    };
    MDBBootstrapModule = __decorate([
        core_1.NgModule({ exports: MODULES })
    ], MDBBootstrapModule);
    return MDBBootstrapModule;
}());
exports.MDBBootstrapModule = MDBBootstrapModule;
