// angular-bootstrap-md/popover/popover.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var common_1 = require("@angular/common");
var component_loader_1 = require("../utils/component-loader");
var positioning_1 = require("../utils/positioning");
var popover_config_1 = require("./popover.config");
var popover_directive_1 = require("./popover.directive");
var popover_container_component_1 = require("./popover-container.component");
var PopoverModule = /** @class */ (function () {
    function PopoverModule() {
    }
    PopoverModule_1 = PopoverModule;
    PopoverModule.forRoot = function () {
        return {
            ngModule: PopoverModule_1,
            providers: [popover_config_1.PopoverConfig, component_loader_1.ComponentLoaderFactory, positioning_1.PositioningService]
        };
    };
    PopoverModule = PopoverModule_1 = __decorate([
        core_1.NgModule({
            imports: [common_1.CommonModule],
            declarations: [popover_directive_1.PopoverDirective, popover_container_component_1.PopoverContainerComponent],
            exports: [popover_directive_1.PopoverDirective],
            entryComponents: [popover_container_component_1.PopoverContainerComponent]
        })
    ], PopoverModule);
    return PopoverModule;
    var PopoverModule_1;
}());
exports.PopoverModule = PopoverModule;
