// angular-bootstrap-md/tooltip/tooltip.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var common_1 = require("@angular/common");
var core_1 = require("@angular/core");
var tooltip_component_1 = require("./tooltip.component");
var tooltip_directive_1 = require("./tooltip.directive");
var tooltip_service_1 = require("./tooltip.service");
var component_loader_1 = require("../utils/component-loader");
var positioning_1 = require("../utils/positioning");
var MDBTooltipModule = /** @class */ (function () {
    function MDBTooltipModule() {
    }
    MDBTooltipModule_1 = MDBTooltipModule;
    MDBTooltipModule.forRoot = function () {
        return {
            ngModule: MDBTooltipModule_1,
            providers: [tooltip_service_1.TooltipConfig, component_loader_1.ComponentLoaderFactory, positioning_1.PositioningService]
        };
    };
    ;
    MDBTooltipModule = MDBTooltipModule_1 = __decorate([
        core_1.NgModule({
            imports: [common_1.CommonModule],
            declarations: [tooltip_directive_1.TooltipDirective, tooltip_component_1.TooltipContainerComponent],
            exports: [tooltip_directive_1.TooltipDirective],
            entryComponents: [tooltip_component_1.TooltipContainerComponent]
        })
    ], MDBTooltipModule);
    return MDBTooltipModule;
    var MDBTooltipModule_1;
}());
exports.MDBTooltipModule = MDBTooltipModule;
