// angular-bootstrap-md/navbars/navlinks.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var navbar_service_1 = require("./navbar.service");
var core_1 = require("@angular/core");
var router_1 = require("@angular/router");
var NavlinksComponent = /** @class */ (function () {
    function NavlinksComponent(_navbarService) {
        this._navbarService = _navbarService;
        this.linkClick = new core_1.EventEmitter();
    }
    NavlinksComponent.prototype.ngAfterContentInit = function () {
        var that = this;
        setTimeout(function () {
            that.links.forEach(function (element) {
                element.nativeElement.onclick = function () {
                    that._navbarService.setNavbarLinkClicks();
                };
            });
        }, 0);
    };
    NavlinksComponent.prototype.ngAfterViewInit = function () {
    };
    __decorate([
        core_1.ContentChildren(router_1.RouterLinkWithHref, { read: core_1.ElementRef, descendants: true }),
        __metadata("design:type", core_1.QueryList)
    ], NavlinksComponent.prototype, "links", void 0);
    __decorate([
        core_1.Output(),
        __metadata("design:type", Object)
    ], NavlinksComponent.prototype, "linkClick", void 0);
    NavlinksComponent = __decorate([
        core_1.Component({
            selector: 'navlinks',
            template: "\n        <ng-content></ng-content>\n    ",
        }),
        __metadata("design:paramtypes", [navbar_service_1.NavbarService])
    ], NavlinksComponent);
    return NavlinksComponent;
}());
exports.NavlinksComponent = NavlinksComponent;
